kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -224,19 +280,42 @@ public void closeConnection() {
      * @param header the response header
      * @param response the response body
      */
-    private void forwardShortCircuitResponse(ResponseHeaderData header, ApiMessage response) {
+    private CompletionStage<Void> forwardShortCircuitResponse(ResponseHeaderData header, ApiMessage response) {
         if (response.apiKey() != decodedFrame.apiKey().id) {
-            throw new AssertionError(
-                    "Attempt to respond with ApiMessage of type " + ApiKeys.forId(response.apiKey()) + " but request is of type " + decodedFrame.apiKey());
+            return CompletableFuture.failedFuture(new AssertionError(
+                    "Attempt to respond with ApiMessage of type " + ApiKeys.forId(response.apiKey()) + " but request is of type " + decodedFrame.apiKey()));
         }
         DecodedResponseFrame<?> responseFrame = new DecodedResponseFrame<>(decodedFrame.apiVersion(), decodedFrame.correlationId(), header, response);
         decodedFrame.transferBuffersTo(responseFrame);
         if (LOGGER.isDebugEnabled()) {
             LOGGER.debug("{}: Forwarding response: {}", channelDescriptor(), decodedFrame);
         }
-        channelContext.fireChannelRead(responseFrame);
-        // required to flush the message back to the client
-        channelContext.fireChannelReadComplete();
+        if (!channelContext.executor().inEventLoop()) {
+            CompletableFuture<Void> future = new CompletableFuture<>();
+            channelContext.executor().execute(new Runnable() {
```

Comment:
This is what is done behind the scenes when you call methods like `fireChannelRead` but you aren't in an event loop thread. I've done this so we can complete a future after we fire the channel read methods within the eventloop thread.